src/generators.cpp

```diff
@@ -96,77 +92,73 @@ OrtEnv& GetOrtEnv() {
   return *GetOrtGlobals()->env_;
 }
 
+// Fallback to copy between two separate device buffers by going through CPU memory (slow unless we're the CPU device)
+void CopyThroughCpu(DeviceBuffer& dest, size_t begin_dest, DeviceBuffer& source, size_t begin_source, size_t size_in_bytes) {
```

Comment:
Should this go inside device interface file as a free function as opposed to generators.cpp?

Reply:
It's used by the cuda provider directly, so having it inside the CPU interface made that difficult. I might change it going forward once we have more shared providers and base class implementations (inheritance doesn't work across shared libraries, so passing in the CPU provider as the base interface might be the solution).